### ISSUE TITLE
Missing example of coroutine as member function

### ISSUE THREAD

**Issue by stephenwhittle** [tzcnt/tmc-examples#108]
Thanks for providing TMC! Haven't used C++20 native coroutines before and I'm enjoying it so far with the additional ergonomics the library provides. 

The documentation at https://www.fleetcode.com/oss/tmc/docs/dev/task.html mentions making the coroutine a class or struct member if it needs to capture `this`:
`If you need to capture *this*, then consider making the coroutine a member function.`

However, I couldn't find a reasonable example of this, especially given that a callable object will suffer from the same issue as a lambda (unsurprising, given that a lambda is a callable object) - namely that capturing `this` is done by value as a pointer, and doesn't keep the surrounding object alive.

Do you have any recommendations as to patterns that can be used with tmc to extend the lifetime of the object? 
There's this sort of thing that I've seen done without tmc:
`tmc::task<Result> operator()(this auto Self)` which leverages explicit object parameters to force capture of the object into the coroutine frame, as it is now a parameter to the coroutine function. 
This has the downside of requiring the coroutine's implementation to be in the header as it is now a template, and traditional alternatives like asio-style capture via shared_ptr to the coroutine's enclosing object feel somewhat icky given the implicit `this` is still there too.

The broader context here is that I have an existing asynchronous SDK that's currently leveraging asio purely for the underlying async runtime (targeting platforms that aren't POSIX compliant, so it's async networking and file io functionality are useless for us).

The codebase has a large number of asynchronous [tasks](https://github.com/modio/modio-sdk/blob/0ccc02c947cf6027c2091a815a90acc98d8f5f8a/modio/modio/detail/ops/auth/AuthenticateUserExternal.h) modelled as callable objects which retain their state, and we've used the switch-based Duffs device mechanism to make the code read in a linear fashion while being able to compose them together via `async_compose`. 
Being able to retain this overall style but switch to native coroutines would substantially improve things, but the lifetime question has me a little stumped thus far, so a demonstration of how you'd solve this idiomatically in the context of your library would be quite helpful. 
For example, does it make sense to somehow derive or associate the underlying object with the tmc::task<> that's initially returned? Those are move-only so would suit places where our task objects own memory via unique_ptr, but as I mentioned, this is the first time I've looked at native coroutines so am not confident in 
a) whether this is feasible 
b) that the lifetime of the task<> object is guaranteed to exceed the duration of the coroutine method itself, even if it is, 
given I'm still learning a lot about the underlying machinery and transformations going on..

Feel free to convert this to a discussion or similar if that's more appropriate.

**Comment by tzcnt** [comment]
Since I originally wrote this, I've learned that it's safe to use a lambda capture as long as you create and await the lambda in the same line, as the lambda object is guaranteed to live to the end of the full-expression:
```cpp
int i = 0;
int j = co_await [i]() -> tmc::task<int> { co_return i + 1; }();
```

Separating the creation and invocation of the lambda into a named variable is OK also (the coroutine holds a reference to this named lambda object):
```cpp
auto func = [i]() -> tmc::task<int> { co_return i + 1; };
int k = co_await func();
```

What's NOT OK is separating the execution of the lambda+creation of the task from awaiting the task:
```cpp
auto task = [i]() -> tmc::task<int> { co_return i + 1; }();
// unsafe! task holds a dangling reference to the lambda object destroyed at the end of the previous expression
co_await std::move(task);
```

This can bite you because it's fragile and can be broken by seemingly obvious refactors:
```cpp
// this works
int i = 0;
std::vector<int> values;
values.push_back(co_await [i]() -> tmc::task<int> { co_return i + 1; }());
values.push_back(co_await [i]() -> tmc::task<int> { co_return i + 2; }());
values.push_back(co_await [i]() -> tmc::task<int> { co_return i + 3; }());

// refactor to use a loop, and also runs in parallel...
// oops, the coroutines now have dangling references to the lambdas
int i = 0;
auto sg = tmc::spawn_group<0, tmc::task<int>>();
for (int v = 1; v <= 3; ++v) {
  sg.add([i, v]() -> tmc::task<int> { co_return i + v; }());
}
std::vector<int> values = co_await std::move(sg);
```

So because the safety of the single-expression, immediately-awaited immediately-invoked lambda is only enforced by convention (you won't get a warning if you make changes that look safe but really break things) I just recommend never using lambdas with captures in coroutines.

**Comment by tzcnt** [comment]
Capturing `this` is essentially the same problem. For example say we have an existing workflow that involves calling `load_file` coroutine followed by `process_file` function, and we want to test forking this off to run in parallel. It might really convenient to quickly whip up a lambda to do this work, but this is incorrect for the same reasons as above.
```cpp
auto forked = tmc::spawn(
  [this]() -> tmc::task<void> {
    auto f = co_await load_file();
    process_file(f);
  }()
).fork();

// ... do stuff while the forked task runs ...
// oops, the lambda went out of scope at the end of the previous expression
co_await std::move(forked);
```

You *could* solve this problem by using a named lambda variable:
```cpp
auto lambda = [this]() -> tmc::task<void> {
  auto f = co_await load_file();
  process_file(f);
}
auto forked = tmc::spawn(lambda()).fork();
// ... do stuff while the forked task runs ...
co_await std::move(forked);
```

But at that point since the declaration has to be separate, I think it's cleaner to just use a member function:
```cpp
tmc::task<void> process_and_load() {
  auto f = co_await load_file();
  process_file(f);
}

// usage
auto forked = tmc::spawn(process_and_load()).fork();
// ... do stuff while the forked task runs ...
co_await std::move(forked);
```

**Comment by tzcnt** [comment]
Regarding member function coroutines, they work just fine as long as you wait for them to finish before the object they are operating on goes out of scope. Forgive the contrived example below, it's just for exposition:
```cpp
struct fileData {};
std::unordered_map<std::string, std::string> loadDictionary();

// type with a member function coroutine
struct reader {
  std::unordered_map<std::string, std::string> dictionary;
  tmc::task<fileData> unpackFile(std::string filename);
};

// usage in some other coroutine
tmc::task<std::vector<fileData>> loadAllFilesInParallel(std::vector<std::string> filenames) {
  reader r;
  r.dictionary = loadDictionary();
  auto fg = tmc::fork_group<fileData>(filenames.size());
  for (auto& f : filenames) {
    fg.fork(r.unpackFile(f));
  }
  // this is completely safe - because we wait for all of the tasks that referenced `r` to finish before it went out of scope
  co_return co_await std::move(fg);
}
```

**Comment by tzcnt** [comment]
So, fork-join is the usual answer to the "lifetime of referenced objects" problem, whether that object is a regular parameter or the implicit `this` of a member function. It's fine to even pass references down several levels into coroutines, as long as you join all the coroutines at each layer of the child call stack before returning.

However if you don't want to do fork-join - that is, you want the lifetime of the coroutine task to outlive the scope where it was created (most easily accomplished in tmc by using the `.detach()` customizer), and it wants to use objects that were created within that scope, you have a few options:

move or copy the object into the coroutine frame by passing it as a value parameter:
```cpp
// declaration - the task accepts `d` by value so it is owned by the coroutine frame
tmc::task<void> long_running_background_task(some_data d);

// usage
tmc::task<void> creates_tasks() {
  some_data d;
  tmc::spawn(long_running_background_task(std::move(d))).detach();
}
```

You could do the same using a `unique_ptr` parameter, or if you need shared lifetime, with a `shared_ptr`. Basically, in this scenario, as long as you avoid any kind of reference or pointer parameter, and take values only, you will be good, as each coroutine owns a copy of the values passed into it within its coroutine frame.

**Comment by tzcnt** [comment]
> For example, does it make sense to somehow derive or associate the underlying object with the tmc::task<> that's initially returned? Those are move-only so would suit places where our task objects own memory via unique_ptr, but as I mentioned, this is the first time I've looked at native coroutines so am not confident in
a) whether this is feasible
b) that the lifetime of the task<> object is guaranteed to exceed the duration of the coroutine method itself, even if it is,
given I'm still learning a lot about the underlying machinery and transformations going on..

I'm not 100% sure on what you mean here but perhaps I can clarify a couple things:
- I won't talk about eager/hot coroutines here because TMC does not offer them. All TMC coroutines are lazy/cold coroutines.
- A lazy coroutine is just a function that returns an object. Calling the coroutine function does nothing other than:
  - Allocate space for a coroutine frame
  - Copy the function parameters into the coroutine frame
  - Return a handle to the coroutine frame (a tmc::task<>)
- The coroutine object only begins executing once it's `resume()` method is called (by submitting it to a thread pool for execution), or if it is directly `co_await`ed by another awaiting coroutine, in which case execution passes to the awaited coroutine by symmetric transfer.

A task<> is just a handle to the coroutine object that lets you call `resume()` or `operator co_await()`. Once the task finishes execution, it will delete itself (the coroutine frame will be freed). The task<> handle could still exist (and point to garbage) but the design of the TMC library aims to prevent this by making task<> move-only so that you can tell a task has been consumed after it is co_awaited. So a task<> is somewhat like a unique_ptr to a coroutine object.

If your task objects own memory that is tied to the lifetime of the task, you could achieve that with a coroutine by either moving the `unique_ptr<T>` into a coroutine function parameter, or simply moving the entire `T` into the coroutine function parameter. The compiler automatically makes the coroutine frame allocation sufficiently sized to accomodate all of the function parameters. So you can actually just place all of your owned allocations directly in the coroutine frame itself via parameter passing.

**Comment by stephenwhittle** [comment]
Thanks for the thorough response. To clarify, with asio, I use this pattern quite heavily at the moment:
```cpp
struct SomeOp{
    auto SomeState;
    auto SomeOtherState;
    void operator()()
    {
        // duff's device state machine in here
    }
};

// entire SomeOp instance stays alive in asio's work queue
asio::post(some_executor,SomeOp());
```
Asio allows me to treat the entire object, rather than just the function, as the coroutine. All of the coroutine's state is kept alive as a result. I guess its *kind of* like manually constructing the coroutine frame myself, as all the data required across suspension points ends up being member data of the struct?

I misinterpreted that line in the docs, as saying there was a way to have member `operator()` contain the coroutine implementation, and passing the entire function object to `spawn()` would result in its lifetime being guaranteed for the duration of the coroutine, in the same way that posting to an `asio::io_context` does, and I had wondered if having my struct actually *be* a `task<T>` might have been a way to facilitate that. 
Your comment about the `task<T>` being functionally just a handle to the coroutine frame rather than somehow *being* the frame/coroutine state itself, was illustrative, thank you.

I'd come across other sources discussing coroutines in general and mentioning the ability to move the outer object into the coroutine as a parameter, so that it was included in the frame. 
Passing the entire function object in *as* the coroutine would have been a cleaner syntax that was closer to how our codebase currently works, so when I saw that line in the docs but no demonstration of it in the samples, I was wondering what neat trick I had missed that would've enabled that ;)

Thanks again

**Comment by tzcnt** [comment]
The typical pattern to replace your struct would be a coroutine with parameters that match what is in your struct (like you said, you were essentially manually constructing a frame).

```cpp
tmc::task<void> SomeOp(auto SomeState, auto SomeOtherState){
  // do stuff with co_await suspend points
}

asio::post(some_executor,SomeOp(state1, state2));
```
